src/Nethermind/Nethermind.Serialization.Rlp/RlpStream.cs

```diff
@@ -676,37 +679,76 @@ public int ReadSequenceLength()
 
         private int DeserializeLength(int lengthOfLength)
         {
-            int result;
-            if (PeekByte() == 0)
+            if (lengthOfLength == 0 || (uint)lengthOfLength > 4)
             {
-                throw new RlpException("Length starts with 0");
+                ThrowArgumentOutOfRangeException(lengthOfLength);
+            }
+
+            // Will use Unsafe.ReadUnaligned as we know the length of the span is same
+            // as what we asked for and then explicitly check lengths, so can skip the
+            // additional bounds checking from BinaryPrimitives.ReadUInt16BigEndian etc
+            ref byte firstElement = ref MemoryMarshal.GetReference(Read(lengthOfLength));
+
+            int result = firstElement;
+            if (result == 0)
+            {
+                ThrowInvalidData();
             }
 
             if (lengthOfLength == 1)
             {
-                result = PeekByte();
+                // Already read above
+                // result = span[0];
             }
             else if (lengthOfLength == 2)
             {
-                result = PeekByte(1) | (PeekByte() << 8);
+                if (BitConverter.IsLittleEndian)
+                {
+                    result = BinaryPrimitives.ReverseEndianness(Unsafe.ReadUnaligned<ushort>(ref firstElement));
+                }
+                else
+                {
+                    result = Unsafe.ReadUnaligned<ushort>(ref firstElement);
+                }
```

Comment:
Would the bytecode be smaller when this would look like:
```c#
result = Unsafe.ReadUnaligned<ushort>(ref firstElement);
if (BitConverter.IsLittleEndian)
{
    result = BinaryPrimitives.ReverseEndianness(result);
}
```

Reply:
Probably would be the same

Reply:
The `if` statement does cause a break in the Jit's reasoning and change the output from the Jit:

Rather than doing a load from memory byte swapped
```asm
G_M000_IG03:
       cmp      esi, 2
       jne      SHORT G_M000_IG04
       movbe    cx, word  ptr [rax]
       movzx    rcx, cx
       jmp      SHORT G_M000_IG06
```
It does a regular load and then calls byte swap on what was loaded
```asm   
G_M000_IG03:
       cmp      esi, 2
       jne      SHORT G_M000_IG04
       movzx    rcx, word  ptr [rax]
       bswap    ecx
       jmp      SHORT G_M000_IG06
```

Reply:
Which is better?

Reply:
One with `movbe` as the CPU can eliminate the `movzx` instruction